viewer/packages/pointclouds/src/potree-three-loader/tree/PointCloudOctreePickerHelper.ts

```diff
@@ -0,0 +1,351 @@
+import {
+  BufferAttribute,
+  Camera,
+  Color,
+  LinearFilter,
+  NearestFilter,
+  NoBlending,
+  Points,
+  Ray,
+  RGBAFormat,
+  Scene,
+  Sphere,
+  Vector3,
+  Vector4,
+  WebGLRenderer,
+  WebGLRenderTarget
+} from 'three';
+import { COLOR_BLACK } from '../rendering/constants';
+import { ClipMode, PointCloudMaterial, PotreePointColorType } from '../rendering';
+import { PointCloudOctree } from './PointCloudOctree';
+import { IPointCloudTreeNode } from './IPointCloudTreeNode';
+import { PickPoint, PointCloudHit } from '../types/types';
+
+export interface RenderedNode {
+  node: IPointCloudTreeNode;
+  octree: PointCloudOctree;
+}
+
+export interface IPickState {
+  renderTarget: WebGLRenderTarget;
+  material: PointCloudMaterial;
+  scene: Scene;
+}
+
+export interface PickParams {
+  pickWindowSize: number;
+  pickOutsideClipRegion: boolean;
+  /**
+   * If provided, the picking will use this pixel position instead of the `Ray` passed to the `pick`
+   * method.
+   */
+  pixelPosition: Vector3;
+  /**
+   * Function which gets called after a picking material has been created and setup and before the
+   * point cloud is rendered into the picking render target. This gives applications a chance to
+   * customize the renderTarget and the material.
+   *
+   * @param material The pick material.
+   * @param renterTarget The render target used for picking.
+   */
+  onBeforePickRender: (material: PointCloudMaterial, renterTarget: WebGLRenderTarget) => void;
+}
+
+/**
+ * Helper class for PointCloudOctreePicker.
+ */
+export class PointCloudOctreePickerHelper {
+  private static readonly helperVec3 = new Vector3();
+  private static readonly helperSphere = new Sphere();
+  private static readonly clearColor = new Color();
+
+  public static prepareRender(
+    renderer: WebGLRenderer,
+    x: number,
+    y: number,
+    pickWndSize: number,
+    pickMaterial: PointCloudMaterial,
+    pickState: IPickState
+  ): void {
+    // Render the intersected nodes onto the pick render target, clipping to a small pick window.
+    renderer.setScissor(x, y, pickWndSize, pickWndSize);
+    renderer.setScissorTest(true);
+    renderer.state.buffers.depth.setTest(pickMaterial.depthTest);
+    renderer.state.buffers.depth.setMask(pickMaterial.depthWrite);
+    renderer.state.setBlending(NoBlending);
+
+    renderer.setRenderTarget(pickState.renderTarget);
+
+    // Save the current clear color and clear the renderer with black color and alpha 0.
+    renderer.getClearColor(this.clearColor);
+    const oldClearAlpha = renderer.getClearAlpha();
+    renderer.setClearColor(COLOR_BLACK, 0);
+    renderer.clear(true, true, true);
+    renderer.setClearColor(this.clearColor, oldClearAlpha);
```

Comment:
Can this be replaced by `WebGLRendererStateHelper` (after adding support for some new fiedls)?